### ISSUE TITLE
[Bug]: Couldn't run DB Migrations from 0.19.3-beta.3 to 0.19.4-beta.4

### ISSUE THREAD

**Issue by ticoombs** [LemmyNet/lemmy#4641]
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [X] Do you agree to follow the rules in our [Code of Conduct](https://join-lemmy.org/docs/code_of_conduct.html)?
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Summary

Upgrading from `-beta.3` to `-beta.4`  results in failed migrations:

```
lemmy-1  | Lemmy v0.19.4-beta.4
lemmy-1  | Error: LemmyError { message: Unknown("Couldn't run DB Migrations: Failed to run 2024-02-24-034523_replaceable-schema with: constraint \"person_aggregates_person_id_fkey\" of relation \"person_aggregates\" does not exist"), inner: Couldn't run DB Migrations: Failed to run 2024-02-24-034523_replaceable-schema with: constraint "person_aggregates_person_id_fkey" of relation "person_aggregates" does not exist, context: SpanTrace [] }
```

### Steps to Reproduce

1. 0.19.3 -> 0.19.4-beta.2
2. 0.19.4-beta.2 -> 0.19.4-beta.3
3. 0.19.4-beta.3 -> 0.19.4-beta.4 - Lemmy will not come online

### Technical Details

Current Indexes:

[indexes.reddthat.log](https://github.com/LemmyNet/lemmy/files/15034065/indexes.reddthat.log)

### Version

0.19.4-beta.4

Related: https://github.com/LemmyNet/lemmy/pull/4333

**Comment by dessalines** [comment]
This is a pretty strange one, because the migration ran fine on both my local machine, and voyager.lemmy.ml 

Grepping for that relation:

```
migrations/2021-03-09-171136_split_user_table_2/down.sql
183:ALTER TABLE user_aggregates RENAME CONSTRAINT person_aggregates_person_id_fkey TO user_aggregates_user_id_fkey;                                        

migrations/2021-03-09-171136_split_user_table_2/up.sql
112:ALTER TABLE person_aggregates RENAME CONSTRAINT user_aggregates_user_id_fkey TO person_aggregates_person_id_fkey;                                                                                                                                                                                                 

migrations/2024-02-24-034523_replaceable-schema/down.sql
1052:    ALTER CONSTRAINT person_aggregates_person_id_fkey NOT DEFERRABLE;                                                                                 

migrations/2024-02-24-034523_replaceable-schema/up.sql                                                                                                     
17:    ALTER CONSTRAINT person_aggregates_person_id_fkey INITIALLY DEFERRED;
```

Type `\d person_aggregates` and show me the output there.

**Comment by dullbananas** [comment]
To clarify, `\d person_aggregates` is typed in `psql`

**Comment by ticoombs** [comment]
'\d person_aggregates'                                                                                       Table "public.person_aggregates"
    Column     |  Type   | Collation | Nullable | Default
---------------+---------+-----------+----------+---------
person_id     | integer |           | not null |
post_count    | bigint  |           | not null | 0
post_score    | bigint  |           | not null | 0
comment_count | bigint  |           | not null | 0
comment_score | bigint  |           | not null | 0
Indexes:
    "person_aggregates_pkey" PRIMARY KEY, btree (person_id)
    "idx_person_aggregates_comment_score" btree (comment_score DESC)
    "idx_person_aggregates_person" btree (person_id)

20 Apr 2024 12:40:21 dullbananas ***@***.***>:

**Comment by dullbananas** [comment]
@ticoombs have you ever ran `diesel migration revert` or `diesel migration redo`, even on very old migrations?

A buggy down.sql file is the only thing I know of that could cause this kind of inconsistency.

**Comment by ticoombs** [comment]
We had a problem with lemmy-webhook. I had to manually get rid of their triggers. It's possible something happened then.

It's probably not a Lemmy specific issue. I might be able to pull a backup from ages ago and see if it still fails.

Do you have any idea on how to fix it?

20 Apr 2024 23:45:54 dullbananas ***@***.***>:

>
> @ticoombs[https://github.com/ticoombs] have you ever ran *diesel migration revert* or *diesel migration redo*, even on very old migrations?
>
> A buggy down.sql file is the only thing I know of that could cause this kind of inconsistency.
>
> —
>

**Comment by dullbananas** [comment]
> Do you have any idea on how to fix it?

If the same migration is still failing, do this to find out what changes to manually do. This must be done with version 0.19.4-beta.4 (the version with the failing migration).

1. Run migrations if you reverted any previous migrations
2. In terminal, run `pg_dump --no-owner --no-privileges --no-table-access-method --schema-only --no-sync -f broken.sqldump`
3. Compare broken.sqldump with this: [schema.zip](https://github.com/LemmyNet/lemmy/files/15052383/schema.zip)

**Comment by dessalines** [comment]
It also might be as simple as adding that index, but I'd be concerned that other areas of your DB has issues if some constraints are missing.

**Comment by dessalines** [comment]
I'm gonna test this with a prod DB to make sure.

**Comment by dessalines** [comment]
Those are the federation test instances, they can get wiped at any time. Although they usually keep a clean upgrade path so we can see if there are any broken migrations.

I'll report back if I get any issues running this migration with the lemmy.ml prod DB locally.

**Comment by dessalines** [comment]
See this: https://stackoverflow.com/a/7623246

It should generate the proper drop function lines.

**Comment by dessalines** [comment]
Whew, glad that's fixed. I'm sure it was a case where a beta / alpha build of ours had the wrong function type, and we corrected it by overwriting an older migration. We try to avoid that bc it breaks a clean upgrade path but sometimes its necessary.

**Comment by dessalines** [comment]
I'll re-open if there are any other issues.

**Comment by ticoombs** [comment]
@dessalines Seems you helped another person but it was certainly a db issue. 

I managed to get a diesel container and found that:

```
  [X] 2024-02-12-211114_add_vote_display_mode_setting
  [X] 2024-02-15-171358_default_instance_sort_type
  [ ] 2024-02-24-034523_replaceable-schema
  [X] 2024-02-27-204628_add_post_alt_text
  [X] 2024-02-28-144211_hide_posts
  [X] 2024-03-06-104706_local_image_user_opt
  [X] 2024-03-06-201637_url_blocklist
  [X] 2024-04-05-153647_alter_vote_display_mode_defaults
  ```

It had deployed half of the migrations some how... :sweat: 

I had to remove all the rows from `person_aggregates` which failed when i added: 
`'alter table only public.person_aggregates add constraint person_aggregates_person_id_fkey foreign key (person_id) references public.person(id) on update cascade on delete cascade;'`

Because the local_image migrate happened too, i had to nuke all items where the id was null: `'delete from local_image where local_user_id is null;'`

I then told diesel to revert the last 5 migrations: `diesel migration revert -n 5` (via a separate container)

Then lemmy would start correctly, perform the needed migrations itself, and we are now on `beta.4`.

Side Note: For some reason lemmy would fail if I ran the migrations manually `diesel migration run`. Technically it would load, and work, but whenever i attempted to comment or post, it failed with a CannotResolvePost. I had to let lemmy perform the migrations otherwise the database would end up in a worse state.

**Comment by maxpoulin64** [comment]
@dessalines I ran into this as well upgrading from 0.19.3 to 0.19.4. I had accidentally pulled 0.19.3beta/rc when the beta was accidentally tagged `:latest` on Docker. From that point on I upgraded through all the betas until the production release, and then stayed on it until the 0.19.4 production release.

The fix was the same as @kroese: I deleted the two `hot_rank` functions and then pasted the entirety of [the down.sql file](https://raw.githubusercontent.com/LemmyNet/lemmy/main/migrations/2024-02-24-034523_replaceable-schema/down.sql) into psql, and then Lemmy started right up and has seeming caught up now.

**Comment by ismailkarsli** [comment]
I had this problem too. However, I did not use the beta version before. I was updating directly from 19.3 to 19.4.

I've removed hot_rank functions and ran only [`CREATE FUNCTION hot_rank`](https://github.com/LemmyNet/lemmy/blob/b2a480f55ce7fd327d46b81f43697069c9befdbd/migrations/2024-02-24-034523_replaceable-schema/down.sql#L1010) part because running whole down file didn't worked.

**Comment by maxpoulin64** [comment]
Update on my end, I ended up with somewhat broken comment threads for some reason. I had to drop the entire `r` schema and rerun the whole contents from `utils.sql` and `triggers.sql` for things to work properly.

Also if anyone's following my steps here, make sure to use psql. Executing the whole script in DBeaver does NOT work correctly.

**Comment by ismailkarsli** [comment]
@maxpoulin64 unpinning posts not working on my instance. I guess I have to do something like you did.

**Comment by maxpoulin64** [comment]
> @maxpoulin64 unpinning posts not working on my instance. I guess I have to do something like you did.

Seems to work on mine. Haven't tested federation of it but I was able to pin and unpin a post just fine. Give it a go, this whole schema is meant to be dropped and replaced every update so drop and replace is safe to do.

**Comment by Showfom** [comment]
Still the same problem when try to upgrade from 0.19.3 to 0.19.4

```
Error: LemmyError { message: Unknown("Couldn't run DB Migrations: Failed to run 2024-02-24-034523_replaceable-schema with: function name \"hot_rank\" is not unique"), inner: Couldn't run DB Migrations: Failed to run 2024-02-24-034523_replaceable-schema with: function name "hot_rank" is not unique, context: SpanTrace [] }
```

**Comment by ismailkarsli** [comment]
@dessalines @Nutomic can you guys please help with this? The problem can be fixed by applying solutions in this thread but pin/unpin issues are persisting. I tried to tweak this as possible but I couldn’t find the solution. I don’t know Lemmy internals a bit.

The fact that both I and @Showfom experienced this while upgrading from 0.19.3 to 0.19.4, makes me think this problem is more widespread.

Please let us know what should we can do to fix this.

**Comment by ismailkarsli** [comment]
> That has nothing to do with the error above, and is fixed by #4837

Really? So re-running replaceable trigger script actually worked and there’s no problem? Cool.

Still, it's annoying that there is a bug while upgrading some specific instances from 19.3 to 19.4 though. Do you think it makes sense to add “remove the hot_rank function from db” to the [steps here](https://github.com/LemmyNet/lemmy-ansible/blob/main/UPGRADING.md#steps)? So new upgrades won’t experience this.

**Comment by ismailkarsli** [comment]
![IMG_1330](https://github.com/LemmyNet/lemmy/assets/17887754/4605f7a8-b46d-412d-8251-ad8cf4a3a857)
